roaring/roaring.go

```diff
@@ -2049,97 +2011,112 @@ func (c *Container) bitmapToRun() {
-func (c *Container) arrayToRun() {
+func (c *Container) arrayToRun(numRuns int32) {
 	statsHit("arrayToRun")
-	c.containerType = containerRun
+	array := c.array()
+	c.typ = containerRun
+	c.mapped = false
 	// return early if empty
 	if c.n == 0 {
-		c.runs = make([]interval16, 0)
-		c.array = nil
-		c.mapped = false
+		c.setRuns(nil)
 		return
 	}
+	if numRuns == 0 {
+		numRuns = arrayCountRuns(array)
+	}
 
-	numRuns := c.arrayCountRuns()
-	c.runs = make([]interval16, 0, numRuns)
-	start := c.array[0]
-	for i, v := range c.array[1:] {
-		if v-c.array[i] > 1 {
+	runs := make([]interval16, 0, numRuns)
+	start := array[0]
+	for i, v := range array[1:] {
+		if v-array[i] > 1 {
 			// if current-previous > 1, one run ends and another begins
-			c.runs = append(c.runs, interval16{start, c.array[i]})
+			runs = append(runs, interval16{start, array[i]})
 			start = v
 		}
 	}
 	// append final run
-	c.runs = append(c.runs, interval16{start, c.array[c.n-1]})
-	c.array = nil
-	c.mapped = false
+	runs = append(runs, interval16{start, array[c.n-1]})
+	c.setRuns(runs)
 }
 
 // runToArray converts from RLE format to array format.
 func (c *Container) runToArray() {
 	statsHit("runToArray")
-	c.containerType = containerArray
-	c.array = make([]uint16, 0, c.n)
+	runs := c.runs()
+	c.typ = containerArray
+	c.mapped = false
 
 	// return early if empty
 	if c.n == 0 {
-		c.runs = nil
-		c.mapped = false
+		c.setArray(nil)
 		return
 	}
 
-	for _, r := range c.runs {
+	array := make([]uint16, c.n)
+	n := int32(0)
+	for _, r := range runs {
 		for v := int(r.start); v <= int(r.last); v++ {
-			c.array = append(c.array, uint16(v))
+			array[n] = uint16(v)
+			n++
 		}
 	}
-	c.runs = nil
-	c.mapped = false
+	if roaringParanoia {
+		if n != c.n {
+			panic("run has fewer bits set than container.n")
+		}
+	}
+	c.setArray(array)
 }
 
 // Clone returns a copy of c.
-func (c *Container) Clone() *Container {
+func (c *Container) Clone() (out *Container) {
 	statsHit("Container/Clone")
-	other := &Container{n: c.n, containerType: c.containerType}
-
-	switch c.containerType {
+	switch c.typ {
 	case containerArray:
 		statsHit("Container/Clone/Array")
-		other.array = make([]uint16, len(c.array))
-		copy(other.array, c.array)
+		cArray := c.array()
+		array := make([]uint16, len(cArray))
+		copy(array, cArray)
+		out = NewContainerArray(array)
 	case containerBitmap:
 		statsHit("Container/Clone/Bitmap")
-		other.bitmap = make([]uint64, len(c.bitmap))
-		copy(other.bitmap, c.bitmap)
+		other := NewContainerBitmap(c.n, nil)
+		copy(other.bitmap(), c.bitmap())
+		out = other
 	case containerRun:
 		statsHit("Container/Clone/Run")
-		other.runs = make([]interval16, len(c.runs))
-		copy(other.runs, c.runs)
+		cRuns := c.runs()
+		runs := make([]interval16, len(cRuns))
+		copy(runs, cRuns)
+		out = NewContainerRun(runs)
 	}
-	return other
+	if roaringParanoia {
+		if out.n != out.count() {
+			panic("cloned container has wrong n")
+		}
+	}
+	// this should probably never happen
```

Comment:
is this comment referring to the previous block?

Reply:
Yes. it should be above it. I think it may originally have referred to a "return nil" because the individual cases returned, but I changed them to create a result to return because I wanted to share the error checking code.